lib/common/validation.js

```diff
@@ -456,6 +460,31 @@ var validators = {
 
     zfs_io_priority: createValidateNumberFn('zfs_io_priority'),
 
+    zfs_snapshot_limit: function (params) {
+        var errs = [];
+
+        if (params.zfs_snapshot_limit === undefined) {
+            // This should make vmadm(1M) not do anything.
+            delete params.zfs_snapshot_limit;
+        } else if (params.zfs_snapshot_limit === null) {
+            // We wanna make sure vmadm(1M) gets it set to '' in this case.
+            params.zfs_snapshot_limit = '';
+        } else {
```

Comment:
I would think if we are setting this to an empty string - it should also take an empty string here (in one of these if/else handlers - otherwise the empty string will end up handled below (passed to the validNumber fn), which I think ends up as NaN... which could be troublesome:
```
> parseInt('', 10);
NaN
```

Reply:
```
[root@moe (kebecloud) ~]# sdc-vmapi /vms/546c8d53-e40e-41fe-8646-8ffa6048ee77?action=update -X POST -d '{"zfs_snapshot_limit": ""}'
HTTP/1.1 409 Conflict
Connection: close
Content-Type: application/json
Content-Length: 161
Date: Wed, 20 May 2020 22:53:01 GMT
Server: VMAPI/9.9.4
x-request-id: 18baaebf-c499-4f8a-92c4-c01e514aa608
x-response-time: 16
x-server-name: bdd3c089-444a-43de-9f0f-ad92dbece8de

{
  "code": "ValidationFailed",
  "message": "Invalid VM update parameters",
  "errors": [
    {
      "field": "",
      "code": "Invalid",
      "message": "Not a valid number: number must be >= 0"
    }
  ]
}
You have new mail in /var/mail/root
[root@moe (kebecloud) ~]# 
```

Reply:
Or is there a difference between "" and '' ?

Reply:
No difference for node/javascript, but the Joyent style guide (and our JS lint tools) says we should only use single quoted ''.

Reply:
So my diff is good (single-quote), and the sample sdc-vmapi invocation seems to catch bad input appropriately.

Reply:
It seems VMAPI will now need to accept '' as a valid value - as you are now setting zfs_snapshot_limit to the empty string:
```
            params.zfs_snapshot_limit = '';
```

So after, when I do a VMAPI get, I will receive a zfs_snapshot_limit with an empty string (right?), and sdc-migrate feeds this back into VMAPI (exactly what it got from VMAPI).